mapie/conformity_scores.py

```diff
@@ -0,0 +1,209 @@
+import sys
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+
+from ._typing import ArrayLike
+
+
+class ConformityScore(metaclass=ABCMeta):
+    """Base class for conformity scores.
+
+    Warning: This class should not be used directly.
+    Use derived classes instead.
+    """
+
+    def __init__(
+        self,
+        sym: bool,
+        consistency_check: bool = True,
+        eps: float = sys.float_info.epsilon,
+    ):
+        """
+        Parameters
+        ----------
+        sym : bool
+            Whether to consider the conformity score as symmetrical or not.
+        consistency_check : bool, optional
+            Whether to check the consistency between the following methods:
+            - get_observed_value and
+            - get_signed_conformity_scores
+            by default True.
+        eps : float, optional
+            Threshold to consider when checking the consistency between the
+            following methods:
+            - get_observed_value and
+            - get_signed_conformity_scores
+            The following equality must be verified:
+            self.get_observed_value(
+                y_pred, self.get_conformity_scores(y, y_pred)
+            ) == y
+            It should be specified if consistency_check==True.
+            by default sys.float_info.epsilon.
+        """
+        self.sym = sym
+        self.eps = eps
+        self.consistency_check = consistency_check
+
+    @abstractmethod
+    def get_signed_conformity_scores(
+        self,
+        y: ArrayLike,
+        y_pred: ArrayLike,
+    ) -> ArrayLike:
+        """Placeholder for get_signed_conformity_scores.
+        Subclasses should implement this method!
+
+        Compute the unsigned conformity scores from the predicted values
+        and the observed ones.
+
+        Parameters
+        ----------
+        y : ArrayLike
+            Observed values.
+        y_pred : ArrayLike
+            Predicted values.
+
+        Returns
+        -------
+        ArrayLike
+            Unsigned conformity scores.
+        """
+
+    def get_conformity_scores(
+        self,
+        y: ArrayLike,
+        y_pred: ArrayLike,
+    ) -> ArrayLike:
+        """Get the conformity score considering the symmetrical property if so.
+
+        Parameters
+        ----------
+        y : ArrayLike
+            Observed values.
+        y_pred : ArrayLike
+            Predicted values.
+
+        Returns
+        -------
+        ArrayLike
+            Conformity scores.
+        """
+        conformity_scores = self.get_signed_conformity_scores(y, y_pred)
+        if self.sym:
+            conformity_scores = np.abs(conformity_scores)
+        return conformity_scores
+
+    @abstractmethod
+    def get_observed_value(
```

Comment:
Although I am not fully in step with the current name get_observed_value, I am not 100% convinced by the new proposal either.

As far as I know, the method in question computes a point of the test distribution from a prediction and a conformity score. Once many points have been obtained, the confidence interval can be computed with the corresponding alpha from this empirical distribution. If so, would get_distribution_obs be a better name?

Reply:
I agree with you on the explanation but I'm not comfortable with "observed" since we are computing the distribution of predicted intervals. What about :
- `get_distribution_uncertainty_bounds`
- `get_distribution_bounds``
- `get_predictioninterval_bound_distribution``
?

Reply:
I don't understand why you would like to have "bound" in the method name? If I understood well, thanks to the method, we get a sample from the prediction distribution and not a sample from the distribution bound (except for the minmax maybe). I have just thought about:
- get_pred_distribution_sample
- get_perturbed_prediction

I had a look a the paper [Predictive inference with the jackknife+](https://arxiv.org/pdf/1905.02928.pdf) and I have not found a name for $\hat{y_{i}}(X)+R^{LOO}_{i}$ yet (except "arrows" in the caption of figure 1: "the boundary is exceeded by a sufficiently small proportion of the two sided arrows").

Reply:
If I understand correctly, we use this method to compute the `lower_bounds` and `upper_bounds` variables in `.predict`. They correspond indeed to the arrows of this figure : https://mapie--135.org.readthedocs.build/en/135/_images/jackknife_jackknife.png. This is why I suggest to call them distributions of bounds.

Reply:
Let's go for `get_estimation_distribution` !

Reply:
Done